### DISCUSSION TITLE
documentation for multiple releases at Jamulus.io

### DISCUSSION THREAD

**Discussion by gene96817** [jamulussoftware/jamulus#1985] [Documentation]
I tried to put this in jamuluswebsite but this is neither a bug nor a translation issue.

As preparation proceeds for 3.8.1, it would be really helpful to have the website preserve documentation for earlier release. Two reasons:
1- Some users choose to delay upgrading, especially the less tech-savvy users. It is sometimes necessary to go back to the documentation for their release to solve problems.

2- When I install a new release, it is helpful to be able to roll back to the old release. Just saving a snapshot of the past installation is not sufficient. There may have been changes to the system too. Or it may be helpful to compare features between the old release and the new release to help with troubleshooting why the new release install is failing. (Sorry my memory is not good enough to remember all the documentation for the old release.)

It would be great if the documentation for the past "few" releases were available. If it is too late to do that, it would be helpful to start this practice of keeping easy access to documentation to previous releases.

**Comment by ann0see** [comment]
Hmm. I think it theory it is possible to have access to the old information since we're using git. If it's easy to implement we can probably develop something. 

Maybe the web archive is a first step: https://web.archive.org/web/20210201010808/https://jamulus.io/

This was tagged around the time of the last release: https://github.com/jamulussoftware/jamuluswebsite/tree/r_website_3_8_0 (but ofc. that's not user friendly at all.)

Edit: I'm afraid I can't develop much in this direction but maybe we find someone else?

**Reply by gene96817** [reply]
Would the web.archive preserve the links for navigation between the pages?
My simple minded idea would be to just push the whole website to be a subdirectory for preserving the site for  3.8.0. Then the top level would evolve for 3.8.1. Would it be too much work to also provide (recover) the frozen site for 3.7.2?

**Reply by ann0see** [reply]
> Would the web.archive preserve the links for navigation between the pages?

I think so, but I'm unsure since web archive doesn't seem to make a snapshot of all pages.

> My simple minded idea would be to just push the whole website to be a subdirectory for preserving the site for 3.8.0. 

Yes, the sub folder solution would work eventually, but needs a lot of effort (links/styling would be broken because of the hacks around translation). Not saying it's impossible though.

**Reply by ann0see** [reply]
I'd still prefer the git approach. In theory, you could even host the old versions yourself:
1. Create a github user account (e.g. jamulusold)
2. Enable the personal github page (this will then be available as jamulusold.github.io): https://pages.github.com/
3. Clone (or fork?) the jamulussoftware/jamuluswebsite repository and add the content to a repo called jamulusold.github.io
4. Checkout the last commit from the previous release (you will want to display this one jamulusold.github.io (especially have a look at the gh-pages branch)): https://www.git-tower.com/learn/git/faq/git-checkout-commits/
5. Enable GitHub pages in the settings of the repo you just created
6. go to jamlusold.github.io to see the old version of the website

Note: 

* some links in the _config.yml file might need to be changed from jamulus.io to jamulusold.github.io in order to allow google to find the correct translations.
* you might need to git push --force the old commit up to github.

**Reply by gene96817** [reply]
The approach you suggest is really a heavy lift for me. I am barely knowledgeable enough to get a server running. And if another even less technical musician needs help, the git approach won't work and I don't have the cycles to do this for more than one very special friend. You forget how super smart you are. What is easy for you is a super-human effort for a lot of people.

**Reply by ann0see** [reply]
I'm not as smart as you might think I am ;-). I'm quite sure you - or someone else - would be able to handle that.

Since you got the server running (which involves CLI), I'm sure you could handle this (and provide it for a larger community).
The basic skills you need are:
1. Some basic knowledge of how to create and use a GitHub account (Steps 1 to 2 and 5 to 6)
2. Yes, this might get tricky: Some knowledge of git/cli (Steps 2-4)

The only problem why I can't do that is that I don't want to maintain another repository.

**Reply by gene96817** [reply]
Believe me... it took me weeks to get my server running properly.
In releasing 3.8.1, why can't the current webserver (mostly a file directory) be cloned and reached by an old-release-3.8.0 link.
Then the main directory (top-level) be for the new release-3.8.1. Then the process repeats for each new release.
Yes, that sounds too easy. What would make that hard to implement?

**Reply by ann0see** [reply]
Mainly the change of the website root and the way we're using relative and absolute links. 

If the content is no longer on jamulus.io but e.g. on jamulus.io/3.8.0 a lot would no longer work as expected.

For example, see:
https://github.com/jamulussoftware/jamuluswebsite/blob/release/_includes/headtags.html

As you see, we're e.g. linking /assets/css/fw.css which is this file: https://github.com/jamulussoftware/jamuluswebsite/blob/release/assets/css/fw.css

As soon as the actual fw.css file from release 3.8.0 is in /3.8.0/assets/css/fw.css we'd suppose that once we access jamulus.io/3.8.0/ also only the /3.8.0/assets/css/fw.css file is loaded. But since we're linking ot /assets/css/fw.css not /3.8.0/assets/css/fw.css this gets a problem (since we're using the wrong file). 
For css this might not be that bad. But as soon as we start with the translation the same issue comes up:
If you look at

https://github.com/jamulussoftware/jamuluswebsite/blob/893bb07bf4d1a9c86d092f20907e3be6eeb16be6/_includes/langselect.html#L11

(this is where most of the language switching takes place), you'll see that this code does the following (in brief):
Goal: Generate a link to all related pages in other languages
Implementation:

Let's assume we're on an English page. The computer still needs to link all other languages in the language switcher. Therefore it goes through English, German, Spanish, etc and inserts a link (here called "langlink").
If it sees that it's currently going through "English" (line 8) we don't prepend the language prefix. 
If it sees that it's currently another language (e.g. "German"), we do prepend the language prefix as follows: /de/[name of the site in english we are currently on] (note the /de/ at the beginning).

As above we have the same problem:

If we're in /3.8.0/wiki/Getting-Started a click on /de/wiki/Getting-Started would link to the Getting-Started page of the up-to-date release, not to /3.8.0/de/wiki/Getting-Started.

Solution: we could use some html magic with the so called base tag to trick the browser into thinking that /3.8.0/ is in fact the root. However after having tried that in the past, it didn't work as I intended and resulted in broken sites.

**Reply by ann0see** [reply]
My answer is not 109% correct, but mainly an illustration of the issue.

If you want, I can try to guide you through my suggestion above via DC/Matrix chat tomorrow. Maybe you could ensure that you have:
1. a pc with git installed 
2. Created a GH account

**Reply by gene96817** [reply]
Oh, I forgot that a few month, after more than a week of trying, I never succeeded in building the documentation from github. (It was so painful I forgot.   LOL)

I wish I can help but since I cannot build the documentation on my MacBook there is nothing I can do.

Here is a deal. If someone will help me get succeed in building the documentation, I will spend at least a week looking at the documentation.

**Reply by gene96817** [reply]
Looking at the long explanation from an hour ago, is this a problem related to relative vs absolute addressing. 
One example was https://github.com/jamulussoftware/jamuluswebsite/blob/release/assets/css/fw.css.
Does this mean that we need to plan ahead so that the releases should be in 
https://github.com/jamulussoftware/jamuluswebsite/blob/release/assets/css/3.8.0/fw.css
and then in https://github.com/jamulussoftware/jamuluswebsite/blob/release/assets/css/3.8.1/fw.css.
In that way the source files are all in a "release-N" tree and then the next release is in a "release-N+1" tree. 

I am not asking this to be retroactive but it could be really worthwile to start doing this going forward from the current 3.8.0 set of files.

**Reply by ann0see** [reply]
Yes, but I think this is much more complicated than the git approach ;-).


The git approach is basically a copy of the site (+ all the hosting and build logic on GitHub). You'd only need to remember a few commands to update the page. GitHub even builds the site for you and publishes it on the web.

Thanks to @ignotus666, you could work locally by using the .zip files which each PR produces and test the site without a full build environment on your mac. But I assume the git approach is simpler and more accessible.

If you have time, you might want to try the local approach, but I'd strongly suggest not (at least for now): https://gist.github.com/ann0see/f738f79fe907487544eb128acf504d4d

What about I create the GitHub account and give you hints on how to handle it?

**Reply by jamulusold** [reply]
Voilà: https://jamulusold.github.io/ (that's a fork of jamulussoftware/jamuluswebsite but with GH pages enabled and the robots.txt file modified so that google doesn't index it)

**Reply by jamulusold** [reply]
Currently this website is up to date, but as soon as the next release comes up, you'd simply not update it. Once the next release comes up, you'd click on "fetch upstream" to "freeze" the site 

![FetchUpstream](https://user-images.githubusercontent.com/89974451/131826846-0d5059ff-c4f9-407d-8164-e916b2ad1d48.png)
(note: you could handle the website via cli and might need to if there are conflicts. However, the I'm sure there are many people in the community who can help you to solve the conflicts).

**Comment by ann0see** [comment]
By the way: if you want to contribute to the documentation, feel free to edit the specific files here: https://github.com/jamulussoftware/jamuluswebsite/tree/next-release/wiki

**Reply by gene96817** [reply]
Thanks for the invite. I'll find some  time to wade through some of the files.

**Comment by ghost** [comment]
I think jamulus 3.6.2 is a good point for older documentation, it took into consideration slower development to this point and Debian releases.

**Comment by ann0see** [comment]
Who wants the login data to @jamulusold ? I'd share them privatly with someone (encrypted E-Mail/messenger/whatever) @gene96817 @DavidSavinkoff ?

Still happy to onboard someone to this approach, but I don't want to maintain it.

**Reply by ghost** [reply]
I'm not sure of what I'll be getting into, so I may take some time deciding on your offer (some learning may be required).

Nevertheless, here is a documentation snapshot for Jamulus 3.6.2
https://github.com/jamulussoftware/jamuluswebsite/tree/ed909699ada9d86464b44d6f8dc9494a2f251acc
If you could reconstruct this documentation and put it online (or make offline browseable html webpages) I think you would cover all older versions reasonably.

**Reply by ann0see** [reply]
It’s not that bad ;-). You‘d just need to checkout and force push old commits to the repo which would then re construct the page.

**Reply by jamulusold** [reply]
Comment moved here: https://github.com/jamulussoftware/jamulus/discussions/1985#discussioncomment-1278653

**Reply by jamulusold** [reply]
The only thing you need to keep in mind is that the robots.txt file should disallow indexing of the page to ensure jamulus.io still doesn't suffer from duplicate content/SEO issues.

**Reply by gene96817** [reply]
I am glad to see this discussion. I apologize for being a dunce. What do I need to know to be able to review (a) the emerging documentation for 3.8.1 and (b) read the mentioned past documentation? I am good at following instructions. As some of you know, as soon as I get an error in the process, I easily get stuck. 

I installed git on my MacBook with "brew install git".  It completed without any error messages. Then from @jamulusold above I ran "git remote add upstream git@github.com:jamulussoftware/jamulus". The results was  "fatal: not a git repository (or any of the parent directories): .git".  So now I am stuck. It worked for @jamulusold, what did I do wrong?

**Reply by ann0see** [reply]
@jamulusold is me ;-). 

What you need to do:
1. `git clone git@github.com:jamulusold/jamulusold.github.io`
2. `cd jamulusold.github.io`
3. the mentioned commands from above. In order to execute `git push --force origin HEAD:gh-pages` I need to give you write access to the repo (and probably the password too). Will do that

**Reply by jamulusold** [reply]
You should have received an invite now and [the instructions should be up to date](https://github.com/jamulussoftware/jamulus/discussions/1985#discussioncomment-1278653) (hopefully)

The site is available at https://jamulusold.github.io

**Reply by jamulusold** [reply]
> What do I need to know to be able to review (a) the emerging documentation for 3.8.1

That's not possible at the moment, but clearly worth a discussion. Could you please open an issue on jamuluswebsite and ping gilgongo and ann0see?

**Comment by jamulusold** [comment]
Actually it's much easier than I thought if we only use the gh-pages branch: https://github.com/jamulussoftware/jamuluswebsite/commits/gh-pages

Once you know the commit (=state of site) on the [gh-pages branch of jamulussoftware/jamuluswebsite](https://github.com/jamulussoftware/jamuluswebsite/commits/gh-pages) which you want to publish, it's easy:

First of all - if you haven't already, clone the repo (=copy to disk). You only need to do this once:
```
git clone git@github.com:jamulusold/jamulusold.github.io
```

Now go into the folder with all the files:
```
cd jamulusold.github.io
```

Now add the jamulussoftware/jamuluswebsite repository as "upstream" repository to ensure getting the latest data which is published on jamulus.io:

```
git remote add upstream git@github.com:jamulussoftware/jamuluswebsite
```

***

**The following needs to be done on every update:**
`git fetch upstream` - to update your local copy to get the latest commits from jamulussoftware/jamuluswebsite
`git checkout <commit ID>` - to get the state you want to publish (you got the commit ID from the link I mentioned above)
`git push --force origin HEAD:gh-pages` - to upload the site to https://jamulusold.github.io

Afterwards run the following commands to ensure Google doesn't index it:

```
git checkout origin/gh-pages && git pull
```

Now edit robots.txt (e.g. with `nano robots.txt`)
...and ensure that it has the following content (this tells Google not to index the page):

```
User-agent: *
Disallow: /
```
Exit nano with `CTRL-X`, now confirm writing to disk with `y` which will exit nano.
Now add robots.txt to the repo with the following commands:
`git add robots.txt`
`git commit -m "Disallow indexing from Google"`
`git push` - to upload the last change to the website.

Voilà, the site should be available here: https://jamulusold.github.io
